/*
 * Wildfire's Female Gender Mod is a female gender mod created for Minecraft.
 * Copyright (C) 2023-present WildfireRomeo
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.wildfire.main.entitydata;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import com.wildfire.main.config.Configuration;
import net.minecraft.core.component.DataComponents;
import net.minecraft.nbt.NbtOps;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.entity.player.PlayerModelPart;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.component.CustomData;
import org.joml.Vector3f;
import org.jspecify.annotations.Nullable;

/// Data component-like class for storing player breast settings on armor equipped onto armor stands
///
/// Note that while this is treated similarly to any other [`data component`][DataComponents] for performance reasons,
/// this is never written as its own component on item stacks, but instead uses the [`custom NBT data component`][DataComponents#CUSTOM_DATA]
/// (under the `WildfireGender` key) for compatibility with vanilla clients on servers.
public record BreastDataComponent(float breastSize, float cleavage, Vector3f offsets, boolean jacket, float breastScale, @Nullable CustomData nbtComponent) {

    private static final String KEY = "WildfireGender";
    private static final Codec<BreastDataComponent> CODEC = RecordCodecBuilder.create(instance -> instance.group(
            Configuration.BUST_SIZE.codec()
                    .optionalFieldOf("BreastSize", 0f)
                    .forGetter(BreastDataComponent::breastSize),
            Configuration.BREASTS_CLEAVAGE.codec()
                    .optionalFieldOf("Cleavage", Configuration.BREASTS_CLEAVAGE.getDefault())
                    .forGetter(BreastDataComponent::cleavage),
            Codec.BOOL
                    .optionalFieldOf("Jacket", true)
                    .forGetter(BreastDataComponent::jacket),
            Configuration.BREASTS_OFFSET_X.codec()
                    .optionalFieldOf("XOffset", 0f)
                    .forGetter(component -> component.offsets.x),
            Configuration.BREASTS_OFFSET_Y.codec()
                    .optionalFieldOf("YOffset", 0f)
                    .forGetter(component -> component.offsets.y),
            Configuration.BREASTS_OFFSET_Z.codec()
                    .optionalFieldOf("ZOffset", 0f)
                    .forGetter(component -> component.offsets.z),
            Configuration.BREASTS_SCALE.codec()
                    .optionalFieldOf("BreastScale", Configuration.BREASTS_SCALE.getDefault())
                    .forGetter(BreastDataComponent::breastScale)
        ).apply(instance, (breastSize, cleavage, jacket, x, y, z, breastScale) -> new BreastDataComponent(breastSize, cleavage, new Vector3f(x, y, z), jacket, breastScale, null))
    );

    public static @Nullable BreastDataComponent fromPlayer(Player player, PlayerConfig config) {
        if(!config.getGender().canHaveBreasts() || !config.showBreastsInArmor()) {
            return null;
        }

        return new BreastDataComponent(config.getBustSize(), config.getBreasts().getCleavage(), config.getBreasts().getOffsets(),
                player.isModelPartShown(PlayerModelPart.JACKET), config.getBreasts().getBreastScale(), null);
    }

    public static @Nullable BreastDataComponent fromComponent(@Nullable CustomData component) {
        if(component == null) {
            return null;
        }

        return CODEC.parse(NbtOps.INSTANCE, component.copyTag().getCompoundOrEmpty(KEY))
                .result()
                .map(breastDataComponent -> breastDataComponent.withComponent(component))
                .orElse(null);
    }

    public void write(ItemStack stack) {
        if(stack.isEmpty()) {
            throw new IllegalArgumentException("The provided ItemStack must not be empty");
        }

        CustomData.update(DataComponents.CUSTOM_DATA, stack, nbt -> nbt.store(KEY, CODEC, this));
    }

    public static void removeFromStack(ItemStack stack) {
        if(stack.isEmpty()) return;
        CustomData component = stack.get(DataComponents.CUSTOM_DATA);
        if(component != null && component.copyTag().contains(KEY)) {
            CustomData.update(DataComponents.CUSTOM_DATA, stack, nbt -> nbt.remove(KEY));
        }
    }

    private BreastDataComponent withComponent(CustomData component) {
        return new BreastDataComponent(breastSize, cleavage, offsets, jacket, breastScale, component);
    }
}
